data/src/main/java/com/microsoft/azure/kusto/data/auth/TokenProviderBase.java

```diff
@@ -0,0 +1,104 @@
+
+import com.microsoft.aad.msal4j.IAccount;
+import com.microsoft.aad.msal4j.IAuthenticationResult;
+import com.microsoft.aad.msal4j.SilentParameters;
+import com.microsoft.azure.kusto.data.exceptions.DataClientException;
+import com.microsoft.azure.kusto.data.exceptions.DataServiceException;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+public abstract class TokenProviderBase {
+    protected static final String ORGANIZATION_URI_SUFFIX = "organizations";
+    protected static final String ERROR_INVALID_AUTHORITY_URL = "Error acquiring ApplicationAccessToken due to invalid Authority URL";
+    protected static final String ERROR_ACQUIRING_APPLICATION_ACCESS_TOKEN = "Error acquiring ApplicationAccessToken";
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+    protected static final int TIMEOUT_MS = 20 * 1000;
+    protected final String clusterUrl;
+    protected final Set<String> scopes;
+    protected final String aadAuthorityUrl;
+
+    TokenProviderBase(@NotNull String clusterUrl, String authorityId) throws URISyntaxException {
+        URI clusterUri = new URI(clusterUrl);
+        this.clusterUrl = String.format("%s://%s", clusterUri.getScheme(), clusterUri.getHost());
+        String scope = String.format("%s/%s", clusterUrl, ".default");
+        scopes = new HashSet<>();
+        scopes.add(scope);
+
+        aadAuthorityUrl = determineAadAuthorityUrl(authorityId);
+    }
+
+    private String determineAadAuthorityUrl(String authorityId) {
+        if (authorityId == null) {
+            authorityId = ORGANIZATION_URI_SUFFIX;
+        }
+
+        String authorityUrl;
+        String aadAuthorityFromEnv = System.getenv("AadAuthorityUri");
+        if (aadAuthorityFromEnv == null) {
+            authorityUrl = String.format("https://login.microsoftonline.com/%s/", authorityId);
+        } else {
+            authorityUrl = String.format("%s%s%s/", aadAuthorityFromEnv, aadAuthorityFromEnv.endsWith("/") ? "" : "/", authorityId);
+        }
+        return authorityUrl;
+    }
+
+    public String acquireAccessToken() throws DataServiceException, DataClientException {
+        IAuthenticationResult accessTokenResult = acquireAccessTokenSilently();
+        if (accessTokenResult == null) {
+            accessTokenResult = acquireNewAccessToken();
+        }
+        return accessTokenResult.accessToken();
+    }
+
+    protected IAuthenticationResult acquireAccessTokenSilently() throws DataServiceException, DataClientException {
+        try {
+            // TODO: MSAL docs say this would throw MsalUiRequiredException, but it doesn't in Java
+            return acquireAccessTokenSilentlyMsal();

```

Comment:
This design is odd.
You make the base class MSAL aware, then override it out where it is not needed.

Reply:
Everything here, from TokenProviderBase and below, is MSAL-aware.

I figured when I did this that it might look strange at first glance, but let's look at it more deeply to understand why I believe it makes more sense.

In all cases where we acquire a token silently, this is the appropriate flow. It's never "overrided out". Now some flows don't acquire tokens silently, in which case this never gets called. And I think it's a design decision whether to get this benefit:
a) in this way in which case some inheritors to whom it's not applicable will get this method, or
b) to add another layer of abstraction so that not only will the flows that don't do silent token retrieval not call this method but also not inherit this method.
I chose (a) because adding another layer of abstraction for only this benefit seemed to make it unnecessarily less maintainable.

By the way, I just want to point out that this method mostly exists to organize exception handling centrally. We want to do this once in the base class, as it's identical in all inheritors.

Reply:
UserToken and Callback do not use MSAL.
MSI would also not use MSAL if and when you support it.

I'd separate this so that so that you have base => MSAL base => MSAL Impls.